File: laundry/wash_wallet_be/resources/js/types/membership_contract.ts
import {
    BaseFilters,
    BaseSortOptions,
    Customer,
    MembershipPlan,
    Outlet,
} from ".";

/**
 * MembershipContract interface
 */
export interface MembershipContract {
    id: number;
    customerId: number;
    membershipPlanId: number;
    startAt: string;
    expiredAt: string;
    status: "active" | "expired" | "replaced" | "cancelled";
    totalPaid: number;
    formattedTotalPaid?: string;
    createdAt: string;
    updatedAt: string;
    customer: Customer;
    outlet: Outlet;
    membershipPlan: MembershipPlan;
    daysRemaining?: number | null;
    discountAmount?: number;
}

/**
 * MembershipContract filter interface extends complete base filters
 */
export interface MembershipContractFilters extends BaseFilters {
    customerId?: number;
    outletId?: number;
    membershipPlanId?: number;
    status?: "active" | "replaced" | "expired";
    startDate?: string;
    endDate?: string;
    expiringIn?: number; // Days
}

/**
 * MembershipContract sort options
 */
export interface MembershipContractSortOptions extends BaseSortOptions {
    column:
        | "id"
        | "startAt"
        | "expiredAt"
        | "totalPaid"
        | "createdAt"
        | "updatedAt"
        | "customerName";
}

/**
 * MembershipContract form data interface
 */
export interface MembershipContractFormData {
    customerId: number;
    outletId: number;
    membershipPlanId: number;
    startAt: string;
    totalPaid: number;
    status?: "active" | "replaced" | "expired";
}

File: laundry/wash_wallet/packages/wash_wallet_domain/lib/src/entities/membership_contract.dart
import 'package:equatable/equatable.dart';
import 'customer.dart';
import 'outlet.dart';
import 'membership_plan.dart';

class MembershipContract extends Equatable {
  final int id;
  final int customerId;
  final int outletId;
  final int membershipPlanId;
  final String? startAt;
  final String? expiredAt;
  final String status;
  final int? replacedById;
  final int? upgradeFromId;
  final double totalPaid;
  final String? formattedTotalPaid;
  final String? createdAt;
  final String? updatedAt;
  final Customer? customer;
  final Outlet? outlet;
  final MembershipPlan? membershipPlan;

  const MembershipContract({
    required this.id,
    required this.customerId,
    required this.outletId,
    required this.membershipPlanId,
    this.startAt,
    this.expiredAt,
    required this.status,
    this.replacedById,
    this.upgradeFromId,
    required this.totalPaid,
    this.formattedTotalPaid,
    this.createdAt,
    this.updatedAt,
    this.customer,
    this.outlet,
    this.membershipPlan,
  });

  factory MembershipContract.fromModel(dynamic model) {
    return MembershipContract(
      id: model.id,
      customerId: model.customerId,
      outletId: model.outletId,
      membershipPlanId: model.membershipPlanId,
      startAt: model.startAt,
      expiredAt: model.expiredAt,
      status: model.status,
      replacedById: model.replacedById,
      upgradeFromId: model.upgradeFromId,
      totalPaid: model.totalPaid,
      formattedTotalPaid: model.formattedTotalPaid,
      createdAt: model.createdAt,
      updatedAt: model.updatedAt,
      customer: model.customer != null
          ? Customer.fromModel(model.customer)
          : null,
      outlet: model.outlet != null ? Outlet.fromModel(model.outlet) : null,
      membershipPlan: model.membershipPlan != null
          ? MembershipPlan.fromModel(model.membershipPlan)
          : null,
    );
  }

  @override
  List<Object?> get props => [
    id,
    customerId,
    outletId,
    membershipPlanId,
    startAt,
    expiredAt,
    status,
    replacedById,
    upgradeFromId,
    totalPaid,
    formattedTotalPaid,
    createdAt,
    updatedAt,
    customer,
    outlet,
    membershipPlan,
  ];
}
